### PR TITLE
BF: Add requirements.txt file to fix pypi package

### LINKED ISSUE #39: Missing version numbers for install dependencies
Installing pyqmix-web from test.pypi.org produces an an error message: 

` ERROR: Could not find a version that satisfies the requirement flask-restplus (from pyqmix-web) (from versions: none)
ERROR: No matching distribution found for flask-restplus (from pyqmix-web)` 

Seems like we need to specify version numbers for all our install dependencies listed in `setup.cfg`?

### PR DESCRIPTION
**Pull request by ArndalAndersen**
Resolves #39

- [x] I have updated `CHANGES.md`

### CODE REVIEW

**Comment by ArndalAndersen** [comment]
The commit was submitted as a 2019.2 release. It's really a shame as the code is _almost_ the same as the 2019.1 release and only serves to fix the pypi packaging. I just thought that this would be the more correct approach instead of hiding the change under the 2019.1 release (where an update was the pypi packaging). Tell me what you think.

**Comment by hoechenberger** [comment]
`requirements.txt` has nothing to do with (PyPI) packaging _per se_. Generally, all dependencies and metadata for the package must be described in `setup.cfg` and `setup.py`. These files contain the information `pip` requires to find out package dependencies. Typically, no version numbers of any dependencies are listed there – except if it's clear that certain versions won't work with your package.

`requirements.txt` is different. Think of it as an equivalent of `conda`'s `environment.yml` for `pip`. Often, but not always, `requirements.txt` will contain version numbers for all listed dependencies. This helps to ensure that you (as a developer) will most likely end up with a working installation if the file is used to install the package including its dependencies via `pip install -r requirements.txt` (the equivalent of `conda env create -f environment.yml`)

There's a lot of confusion around these topics, see e.g. https://caremad.io/posts/2013/07/setup-vs-requirement/
tl;dr: Python packaging is shit and super annoying.

That said, I'm not sure where exactly the issue you observed arises from; but I am quite confident that adding a `requirements.txt` file is not the proper fix.

**Comment by hoechenberger** [comment]
Couldn't reproduce your issue:
```
$ conda create -n pyqmix-web-test python=3.7
$ conda activate pyqmix-web-test
$ pip install .
```
retrieves all deps and installs flawlessly. Could you provide more information on your setup / how to reproduce the error?

**Comment by ArndalAndersen** [comment]
Yes, sure thing (and thanks for the detailed explanation - it did clear things up a bit for me). So, what I did was to package up our code following our own [documentation](https://github.com/psyfood/pyqmix/blob/master/PACKAGING.md). For rigor, I then wanted to test whether I could install the package from `test.pypi.org` before I went ahead and uploaded the package to `pypi.org`, and this is when my issues started. What I did:

`$ conda create -n pyqmix-web-test python=3.7`
`$ conda activate pyqmix-web-test`
`$ pip install -i https://test.pypi.org/simple/ pyqmix-web==2019.1`

Here I would get the error I posted in issue #39. 
I then fixed the issue by including a `requirements.txt`-file and submitted this as the `2019.1.2`-version. In a new environment I could now install `pyqmix-web` by saying:
`$ pip install -i https://test.pypi.org/simple/ pyqmix-web==2019.1.2`
So I thought I had fixed the issue? 

And now I need to correct myself a bit because the story is a _bit_ more complicated than that. Since `flask-restplus` does not exist on `test.pypi.org` the installation would actually fail when I tried the `$ pip install -i https://test.pypi.org/simple/ pyqmix-web==2019.1.2`-command. But it would only fail when trying to install the `flask-restplus` dependency, and no longer fail when installing the `flask`-dependency. So, I was actually happy ;) 

Does this help explain where the error arose from?

**Comment by hoechenberger** [comment]
> Does this help explain where the error arose from?

Yes.

The issue here is in fact that you're pointing `pip` to `test.pypi.org` to not only install `pyqmix-web`, but also _all of its dependencies_. However, we cannot expect all deps to be present on the test server. In fact, we're just in luck that _any_ dependencies other than `pyqmix` are to be found there: as far as I know, the test uploads are more-or-less regularly deleted. So, the test server is really just for trying out whether the upload itself would work, and to maybe have a look at the package description (in our case, the one displayed at https://test.pypi.org/project/pyqmix-web/) Installation with your above command fails for me with both `2019.1` and `2019.1.2` due to the missing `flask-restplus` dependency on the test server.

To test _package installation based on the source distribution you (are going to) upload to PyPI_, simply install that package locally – preferably into a new, clean environment:
```
$ conda create -n pyqmix-web-test python=3.7 --yes && conda activate pyqmix-web-test
$ python setup.py sdist
$ pip install dist/pyqmix_web-2019.1.tar.gz
```
(the `tar.gz` file would have a `.zip` extension on Windows, btw.)
Same procedure would work with the wheel created via `python setup.py bdist_wheel`.

`pip` entirely ignores the `requirements.txt` file here. In fact, this file is not even packaged into the source distribution (sdist), so even if `pip` wanted to, it couldn't honor it during package installation:
```
$ mkdir /tmp/download
$ cd /tmp/download
$ curl -o pyqmix_web.tar.gz https://test-files.pythonhosted.org/packages/b3/06/35d6d08e799ca0fab03616077ba78eff4f1cdb8f2b5ebf0460e0c1817c09/pyqmix_web-2019.1.2.tar.gz  # Got this URL from the PyPI test website
$ tar xzvf pyqmix_web.tar.gz
$ cd pyqmix_web-2019.1.2
$ ls -l
total 264
-rw-r--r--  1 hoechenberger  wheel    112 Mar 25 17:02 AUTHORS.md
-rw-r--r--  1 hoechenberger  wheel    882 May  2 22:56 CHANGES.md
-rw-r--r--  1 hoechenberger  wheel  35147 Mar 25 17:02 LICENSE
-rw-r--r--  1 hoechenberger  wheel    157 Apr 29 11:47 MANIFEST.in
-rw-r--r--  1 hoechenberger  wheel   2014 May  4 14:06 PKG-INFO
-rw-r--r--  1 hoechenberger  wheel   3028 Apr 29 11:47 README.md
drwxr-xr-x  6 hoechenberger  wheel    192 May  4 14:06 pyqmix_backend
drwxr-xr-x  4 hoechenberger  wheel    128 May  4 14:06 pyqmix_frontend
drwxr-xr-x  8 hoechenberger  wheel    256 May  4 14:06 pyqmix_web.egg-info
-rw-r--r--  1 hoechenberger  wheel   1124 May  4 14:06 setup.cfg
-rw-r--r--  1 hoechenberger  wheel    653 Apr 29 11:47 setup.py
-rw-r--r--  1 hoechenberger  wheel  68611 Apr 29 11:47 versioneer.py
```
This is because to add "additional" files to the sdist, we'd those files must be listed in `MANIFEST.in`. Typically, `requirement.txt` files are _not_ included in the sdist as their purpose is… different.

tl;dr: You did everything correctly; the PyPI test server just doesn't contain all our deps, which is fine. Test the package installation locally, and if that works and the project description looks okay on the PyPI test server, you can publish the package on PyPI.

**Comment by ArndalAndersen** [comment]
Well, you do make quite a convincing argument there. So to move on I'll just go ahead and upload pyqmix-web to pypi.

Only one very last (hopefully quick and easy) question. You say that:
> (the tar.gz file would have a .zip extension on Windows, btw.)

But it seems not to matter whether I upload a `.zip` or `.tar.gz` to `pypi.org`, is that correctly understood?

**Comment by hoechenberger** [comment]
> But it seems not to matter whether I upload a `.zip` or `.tar.gz` to `pypi.org`, is that correctly understood?

I'm not entirely sure whether `pip` on Windows can handle `.tar.gz` files; the default `sdist` on Windows is a `.zip` file. You could try to copy the `.tar.gz` to your Windows computer and see whether `pip` can install it?

**Comment by ArndalAndersen** [comment]
I will admit defeat and delete this pull request. 

One very last key learning that I can share: 
On Windows one also creates `.tar.gz` files during packaging. No `.pip` files are created. This also means that Windows can install these `.tar.gz` files just fine. So, the key learning really is that the operating system does not matter when we pack up our software. It's only when we create the standalone that we need a Windows machine.

**Comment by hoechenberger** [comment]
> So, the key learning really is that the operating system does not matter when we pack up our software. It's only when we create the standalone that we need a Windows machine.

Interesting. Because according to the docs, on Windows a zip file should be created by default, and tar.gz on Unix.

There is actually an option when can set in `setup.cfg` to specify which archive type shall be produced: https://gist.github.com/althonos/6914b896789d3f2078d1e6237642c35c#file-setup-cfg-L62
